kmod/src/forest.c

```diff
@@ -662,7 +687,131 @@ void scoutfs_forest_destroy(struct super_block *sb)
 
 	if (finf) {
 		scoutfs_block_put(sb, finf->srch_bl);
+		scoutfs_block_put(sb, finf->oino_bl);
+		vfree(finf->oino_counted_bloom);
 		kfree(finf);
 		sbi->forest_info = NULL;
 	}
 }
+
+/*
+ * Track current open inodes in a counted bloom filter. Save a much smaller
+ * classic BF of open inodes in each transaction, allowing orphan processing to
+ * know when an orphan inode is no longer open on any node.
+ *
+ * Let's call this code: open inode tracking (OINO)
+ */
+
+void scoutfs_forest_oino_opened(struct super_block *sb, struct inode *inode)
+{
+	struct scoutfs_inode_info *si = SCOUTFS_I(inode);
+	DECLARE_FOREST_INFO(sb, finf);
+	u64 hash;
+	int val;
+	int i;
+
+	val = atomic_inc_return(&finf->oino_open_count);
+	WARN_ON_ONCE(val <= 0);
+
+	hash = scoutfs_hash64(&si->ino, sizeof(si->ino));
+	for (i = 0; i < SCOUTFS_FOREST_BLOOM_NRS; i++) {
+		u32 idx = (u32)hash % SCOUTFS_FOREST_BLOOM_BITS;
+		val = atomic_inc_return(&finf->oino_counted_bloom[idx]);
+		WARN_ON_ONCE(val < 0);
+
+		hash >>= SCOUTFS_FOREST_BLOOM_FUNC_BITS;
+	}
+}
+
+void scoutfs_forest_oino_released(struct super_block *sb, struct inode *inode)
+{
+	struct scoutfs_inode_info *si = SCOUTFS_I(inode);
+	DECLARE_FOREST_INFO(sb, finf);
+	u64 hash;
+	int val;
+	int i;
+
+	val = atomic_dec_return(&finf->oino_open_count);
+	WARN_ON_ONCE(val < 0);
+
+	hash = scoutfs_hash64(&si->ino, sizeof(si->ino));
+	for (i = 0; i < SCOUTFS_FOREST_BLOOM_NRS; i++) {
+		u32 idx = (u32)hash % SCOUTFS_FOREST_BLOOM_BITS;
+		val = atomic_dec_return(&finf->oino_counted_bloom[idx]);
+		WARN_ON_ONCE(val < 0);
+
+		hash >>= SCOUTFS_FOREST_BLOOM_FUNC_BITS;
+	}
+}
+
+/*
+ * At the start of the transaction, alloc and mark dirty a meta block in
+ * anticipation of updating the OINO BF at the end of the transaction.
+ */
+int scoutfs_forest_oino_alloc(struct super_block *sb, u64 trans_seq)
+{
+	struct scoutfs_block *new_bl = NULL;
+	struct scoutfs_bloom_block *bb;
+	struct scoutfs_btree_ref *ref;
+	DECLARE_FOREST_INFO(sb, finf);
+	u64 blkno;
+	int ret;
+	int err;
+
+	WARN_ON_ONCE(finf->oino_bl);
+
+	ret = scoutfs_alloc_meta(sb, finf->alloc, finf->wri, &blkno);
+	if (ret < 0)
+		goto out;
+
+	new_bl = scoutfs_block_create(sb, blkno);
+	if (IS_ERR(new_bl)) {
+		err = scoutfs_free_meta(sb, finf->alloc, finf->wri, blkno);
+		BUG_ON(err); /* could have dirtied */
+		ret = PTR_ERR(new_bl);
+		goto out;
+	}
+
+	memset(new_bl->data, 0, SCOUTFS_BLOCK_LG_SIZE);
+
+	scoutfs_block_writer_mark_dirty(sb, finf->wri, new_bl);
+
+	bb = new_bl->data;
+
+	/* use different magic? */
+	bb->hdr.magic = cpu_to_le32(SCOUTFS_BLOCK_MAGIC_BLOOM);
+	bb->hdr.fsid = SCOUTFS_SB(sb)->super.hdr.fsid;
+	bb->hdr.blkno = cpu_to_le64(blkno);
+	bb->hdr.seq = cpu_to_le64(trans_seq);
+
+	ref = &finf->our_log.oino_bloom_ref;
```

Comment:
ok, yes.

Reply:
The first thing I tried -- call scoutfs_free_meta(blkno) if old oino_bloom_ref was nonzero -- did not work: server error emptying freed (-EINVAL) because of overlap detected in scoutfs_ext_insert().

If this isn't working because the block is from a previous transaction, how does one free said block?

Reply:
So what's happening here is that the client is creating a new dirty version of the oino bloom block to write and be applied by the server as the new version referenced by the log_trees item that everyone will read and use.  When the server applies the client's transaction it will be overwriting the existing ref, so part of applying the transaction needs to also free the old blkno.   That's what freeing the old non-zero ref blkno in the client does as it dirties the new oin bloom block -- it adds the old ref blkno to the freed alloc list block that will be given to the server as the transaction is applied.  Dunno why you'd be seeing double frees (what the overlap is finding), we'd have to see the code to tell for sure.  It'd want to look like the freeing of the old forest item bloom block in _bloom_set_bits().

Reply:
duh, I'm freeing it twice for some reason is why :facepalm: